### ISSUE TITLE
Add support for zeroonfig discovery

### ISSUE THREAD

**Issue by DrTree** [MindrustUK/Heatmiser-for-home-assistant#50]
The neohub advertises itself as a zeroconfig device.

- Support manual addition of hub if mdns is not available
- Do not trust the advertised port on the zerocofig enry
- Use the mdns hostname rather than IP when discovered this way

**Comment by stikonas** [comment]
I guess this is now implemented and can be closed?

Personally I had hit a small problem getting this to work but that's unrelated to this integration. I run Home Assistant in rootless podman container but so far I was not able to configure mDNS address resolution inside the container.

**Comment by andriuss-xilinx** [comment]
I was looking a bit at my zeroconf support inside container and it seems to be a common issue.

https://github.com/home-assistant/core/issues/38475

Docker container is alpine based which fails to resolve IP via mdns. If I understand correctly they suggest that integrations resolve hostname->IP themselves.

**Comment by stikonas** [comment]
I was looking a bit at my zeroconf support inside container and it seems to be a common issue.

https://github.com/home-assistant/core/issues/38475

Docker container is alpine based which fails to resolve IP via mdns. If I understand correctly they suggest that integrations resolve hostname->IP themselves.

**Comment by stikonas** [comment]
On second thought I looked a bit more into this and I don't think I myself agree that each integration has to handle hostname->IP resolution.

This only affects docker containers and better way to workaround it is to use mdns->dns proxy (systemd-resolved can do that) on the host instead of adding same code to ALL zeroconf integrations...

So I again think that this issue can be closed.

**Comment by stikonas** [comment]
ping? Shall we close this?

**Comment by DrTree** [comment]
Yes, but... When I reinstalled the integration it didn't pick up the hub automatically so has potentially regressed with API change (although I struggle to see how).  May need a little investigation.

**Comment by MindrustUK** [comment]
Zeroconfig is actually only supported on the NeoHub 2 I believe. I know Heatmiser were doing "stuff" I'm not at liberty to discuss. I've reached out to my contact there for an update and will update this ticket accordingly. I believe this issue should abate with firmware maturity.

**Comment by MindrustUK** [comment]
Heatmiser support discovery of NeoHub2 devices via ‘hubseek’ (see API docs), enable this feature to make 'out of the box' configuration nicer.

**Comment by MindrustUK** [comment]
Needs updates to Home Assistant config flow and underlying UDP sending and receiving in neohubapi to support this.

**Comment by MindrustUK** [comment]
This is a two part solution:
1) Expand the neohubapi to handle discovery elements (Sample code below).
2) Adjust the integration's initial setup config flow to have an option to call optional discovery of NeoHub V2+ devices early in the initial setup process. I will open another task to deal with the configuration and options flow re-writes as a reference to where this should fit in the process.


Sample code that is working and could be modified for inclusion in neohubapi follows:
```
import asyncio
import socket
import json
import logging
from typing import Set, Dict, Tuple

# Configuration for the UDP discovery
DISCOVERY_PORT = 19790  # Heatmiser Discovery port
DISCOVERY_COMMAND = "hubseek"
LISTEN_TIMEOUT = 30  # Time in seconds to listen for responses
BROADCAST_INTERVAL = 10  # Delay in seconds between discovery broadcasts
NUM_BROADCASTS = 3  # Number of broadcasts

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')


# Function to send discovery command over UDP
async def send_discovery_command():
    broadcast_address = ('255.255.255.255', DISCOVERY_PORT)
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
        for _ in range(NUM_BROADCASTS):
            logging.info(f"Sending discovery command: {DISCOVERY_COMMAND}")
            sock.sendto(DISCOVERY_COMMAND.encode('utf-8'), broadcast_address)
            await asyncio.sleep(BROADCAST_INTERVAL)


# Function to listen for UDP responses
async def listen_for_discoveries() -> Set[Dict[str, str]]:
    discovered_devices = set()
    loop = asyncio.get_event_loop()

    # Set up UDP socket
    with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
        sock.bind(('', DISCOVERY_PORT))
        sock.setblocking(False)

        start_time = loop.time()
        while loop.time() - start_time < LISTEN_TIMEOUT:
            try:
                # Use recvfrom to capture both data and address
                data, addr = await loop.run_in_executor(None, sock.recvfrom, 1024)
                device_info = data.decode('utf-8')

                # Parse and validate the device info
                try:
                    device_data = json.loads(device_info)
                    # Check that both "ip" and "device_id" keys are present
                    if {"ip", "device_id"} <= device_data.keys():
                        # Convert to a tuple to ensure uniqueness (e.g., (ip, device_id))
                        device_tuple = (device_data["ip"], device_data["device_id"])
                        if device_tuple not in discovered_devices:
                            discovered_devices.add(device_tuple)
                            logging.info(f"Added unique device: {device_data}")
                    else:
                        logging.warning(f"Invalid device format: {device_info}")
                except json.JSONDecodeError:
                    logging.warning(f"Non-JSON data received and ignored: {device_info}")

            except BlockingIOError:
                await asyncio.sleep(0.1)  # Small delay to avoid busy-waiting

    # Convert set of tuples back to a set of dictionaries for the return format
    return [{"ip": ip, "device_id": device_id} for ip, device_id in discovered_devices]


# Main async function to run discovery
async def discover_devices():
    listener_task = asyncio.create_task(listen_for_discoveries())
    broadcast_task = asyncio.create_task(send_discovery_command())

    await broadcast_task  # Wait for the broadcasting to complete
    devices = await listener_task  # Wait for the listening to complete

    return devices


# Function to call the discovery process
def discover():
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:  # No running event loop, create a new one
        return asyncio.run(discover_devices())
    else:
        return loop.run_until_complete(discover_devices())


if __name__ == "__main__":
    devices = discover()
    logging.info(f"Discovered devices: {devices}")

```

**Comment by stikonas** [comment]
> Sample code that is working and could be modified for inclusion in neohubapi follows:

This doesn't work on the neohub that I'm using :(. Maybe my firmware is too old.
```
2024-12-03 22:51:28,540 - INFO - Sending discovery command: hubseek
2024-12-03 22:51:28,641 - WARNING - Non-JSON data received and ignored: hubseek
```

**Comment by MindrustUK** [comment]
@stikonas Are you still on a NeoHub V1? I think this is only offered on the V2 or Mini, there's a lot of stuff the V1 Hub doesn't support like sensors etc. Not sure why, I can only guess hardware limitations like not enough flash memory or something.

**Comment by ocrease** [comment]
@MindrustUK I've just released v3.2.0-beta.1 which implements discovery. It doesn't use the code above, instead I used one of the core home assistant integrations for inspiration (the elkm1 integration). 

@stikonas, sounds like you have an older hub? If you can try it out and give some feedback that would be great.

**Comment by stikonas** [comment]
@ocrease I can test it but let me check your code first to understand what is expected to happen.

I also have my neohub and home assistant on different subnets, so I need to even see if discovery is even expected to work on my network...

**Comment by stikonas** [comment]
@ocrease I think discovery works.

`Heatmiser Neo Climate` appears in the list of discovered devices. Then I can click add and pick Legacy method and re-add my Hub.

**Comment by stikonas** [comment]
Hmm, one thing I noticed is that it keeps saying New devices are discovered even after I added integration...

**Comment by ocrease** [comment]
@stikonas its supposed to handle that. What happens if you try to add it again, either via the new discovery or via a manual entry? It's supposed to say already configured, but I suspect in your case it will add it again? 

The hub you already configured, is it using an IP address or did you add it via manual entry using a hostname?

I'll need to take a look, but it might not be today

**Comment by stikonas** [comment]
It's fine, no rush here. It said it was already configured, so doesn't add any duplicate devices and then that new discovery button disappears (until the next restart of home assistant).

**Comment by ocrease** [comment]
@stikonas I had a look at the diagnostics output you uploaded the other day and it looks like you had set up your hub using zeroconf discovery. I imagine you have since deleted and re-added it by now, but maybe the integration is reacting to botha zeroconf discovery and a `hubseek` discovery. In an effort to try to narrow it down, could you try updating the `manifest.json` and remove the `zeroconf` line?

**Comment by stikonas** [comment]
Ok, now I've removed `zeroconf` line, it doesn't discover it at all (which is perhaps expected in my network since my HA is on `192.168.1.1/24` and Neohub is on `192.168.2.1/24` (disconnected from the internet).

**Comment by ocrease** [comment]
OK, I think that means I've broken something with `zeroconf`. `hubseek` discovery won't work in your setup because it relies on a UDP broadcast which won't traverse across networks/subnets (I use opnsense and I can use a upd broadcast relay plugin to support that, but it's not supported out of the box)

I have never seen my hub publish a zeroconf message. Do you have homekit enabled on the hub? I don't, and that's the only thing I can think of that would enable it. 

I'll review my change to see what I did that broke zeroconf discovery. I have a few ideas but it's a bit tricky when I can't test it!

**Comment by stikonas** [comment]
> I have never seen my hub publish a zeroconf message. Do you have homekit enabled on the hub? I don't, and that's the only thing I can think of that would enable it.

I don't think so, at least I'm not aware of how I could have enabled homekit. I haven't used the hub outside home assistant and neohubapi. Even my thermostats were paired manually using heatmiser API commands.

In the meantime I'll try to test hubseek with a new test installation of home assistant on the same subnet where neohub is.

**Comment by stikonas** [comment]
No, I don't think my hub supports `hubseek` at all...

`echo -n "hubseek" | nc -b -u 255.255.255.255 19790` does not result in anything even on the same network.

**Comment by ocrease** [comment]
Unfortunately that command is half the story. I struggled to get it to work. That command only publishes the `hubseek` message, it doesn't listen for a response! I had to set up a listener on a different terminal to see the response. I can't remember the exact command, something like 

`nc -u -l -k 192.168.2.255 19790`

It might need a `-b` as well, I don't have my computer to hand to check. Run that on one terminal first, then from another do `echo -n "hubseek" | nc -b -u 255.255.255.255 19790`

**Comment by stikonas** [comment]
Didn't have much luck with `nc` listening mode either. Then I fired `wireshark` just to double check, I indeed didn't get any reply... Still not reply either.

Then I tried sending hubseek to Neohub IP directly (rather than broadcast address) and got ICPM Port unreachable packet back.

So almost surely old Neohub Firmwares don't support hubseek. Perhaps it appeared at around the same time websocket mode was introduced?

**Comment by MindrustUK** [comment]
Hi @stikonas @ocrease 

My initial commands for discovering the device were as follows (Run these in two terminals, one to send one to receive);

`echo -n "hubseek" | nc -b -u 255.255.255.255 19790`

`netcat -ulk -p 19790`

That said, there are three versions of the Heatmiser NeoHub
Neohub Gen 1
Neohub Gen 2
NeoHub Mini

@stikonas as above, I believe your hardware is a Gen 1. An easy way to tell is the Gen 1 is encased in white plastic, the Gen 2 is a black enclosure.

From what I understand there's a hardware limitation on the Gen 1 devices which means they don't support token based (non legacy API) access or discovery.

Can you confirm you are on a Gen 2 Neohub?

@ocrease everything in the Beta is working for me ATM on my NeoHub Gen 2. I'll break out the Gen 1 just to confirm my assumptions above when I get a minute over the next two weeks. I also now have access to a NeoHub Mini, I'll take a machine to where that's located and give it a test when I get the chance.

**Comment by stikonas** [comment]
I'm on Gen 2 NeoHub in black enclosure (though on fairly old firmware, probably 3 years old or so).

**Comment by MindrustUK** [comment]
@stikonas  Great, thanks for confirming. That rules out that theory (Gen1 vs Gen2). Your assertion of Firmware is sounding the most likely.

**Comment by stikonas** [comment]
I've mostly kept it on old version cause otherwise I need to use their app to set up API tokens...

I wonder how they add those to device. Perhaps they have some undocumented API over which they set up API token on Neohub...

**Comment by MindrustUK** [comment]
I've wondered the same, it would be nice not to have to use the app. It's probably some horrible OTA call back to Heatmiser's servers :(

**Comment by ocrease** [comment]
@stikonas The Zeroconf issue you reported with duplicate discoveries should be fixed now in v3.2.0-beta.2

**Comment by stikonas** [comment]
> [@stikonas](https://github.com/stikonas) The Zeroconf issue you reported with duplicate discoveries should be fixed now in v3.2.0-beta.2

Thanks, it seems to be working, so far I don't see any duplicate discoveries.

**Comment by ocrease** [comment]
@MindrustUK @stikonas anything else left to do here?

**Comment by stikonas** [comment]
Nothing from my point of view. I think we can close this. @DrTree are you happy with the current state?

**Comment by MindrustUK** [comment]
@ocrease Think we can call this done too. Closing.

**Comment by ocrease** [comment]
@MindrustUK thanks. I just released v3.2.1 to officially release this (promoted from beta release)
